ClientMapPartitionLostListenerTest.test_mapPartitionLostListener_invoked_fromOtherNode

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest$1.run(ClientMapPartitionLostListenerTest.java:103)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:835)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:849)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.assertMapPartitionLostEventEventually(ClientMapPartitionLostListenerTest.java:97)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.test_mapPartitionLostListener_invoked_fromOtherNode(ClientMapPartitionLostListenerTest.java:137)
```

It could be related to recent changes in the mock network? #7036

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest$1.run(ClientMapPartitionLostListenerTest.java:104)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:737)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:751)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.assertMapPartitionLostEventEventually(ClientMapPartitionLostListenerTest.java:98)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.test_mapPartitionLostListener_invoked_fromOtherNode(ClientMapPartitionLostListenerTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance/com.hazelcast$hazelcast-client-new/1124/testReport/junit/com.hazelcast.client.map/ClientMapPartitionLostListenerTest/test_mapPartitionLostListener_invoked_fromOtherNode/

Build failure: 

https://hazelcast-l337.ci.cloudbees.com/job/Hazelcast-3.maintenance-issue-7049/8/

```
Stacktrace

java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest$1.run(ClientMapPartitionLostListenerTest.java:105)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:737)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:751)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.assertMapPartitionLostEventEventually(ClientMapPartitionLostListenerTest.java:99)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.test_mapPartitionLostListener_invoked_fromOtherNode(ClientMapPartitionLostListenerTest.java:140)
```

```
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest$1.run(ClientMapPartitionLostListenerTest.java:104)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:737)
	at com.hazelcast.test.HazelcastTestSupport.assertTrueEventually(HazelcastTestSupport.java:751)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.assertMapPartitionLostEventEventually(ClientMapPartitionLostListenerTest.java:98)
	at com.hazelcast.client.map.ClientMapPartitionLostListenerTest.test_mapPartitionLostListener_invoked_fromOtherNode(ClientMapPartitionLostListenerTest.java:139)
```

https://hazelcast-l337.ci.cloudbees.com/view/Hazelcast%20Maintenance/job/Hazelcast-3.maintenance-IbmJDK1.8/com.hazelcast$hazelcast-client/26/testReport/junit/com.hazelcast.client.map/ClientMapPartitionLostListenerTest/test_mapPartitionLostListener_invoked_fromOtherNode/

--- a/hazelcast-client-new/src/test/java/com/hazelcast/client/map/ClientMapPartitionLostListenerTest.java
+++ b/hazelcast-client-new/src/test/java/com/hazelcast/client/map/ClientMapPartitionLostListenerTest.java
@@ -7,6 +7,7 @@ import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.map.MapPartitionLostEvent;
 import com.hazelcast.map.MapPartitionLostListenerStressTest.EventCollectingMapPartitionLostListener;
+import com.hazelcast.map.impl.MapPartitionLostEventFilter;
 import com.hazelcast.map.impl.MapService;
 import com.hazelcast.map.listener.MapPartitionLostListener;
 import com.hazelcast.nio.Address;
@@ -21,7 +22,6 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.Collection;
 import java.util.List;
 
 import static com.hazelcast.client.impl.ClientTestUtil.getHazelcastClientInstanceImpl;
@@ -29,11 +29,14 @@ import static com.hazelcast.map.impl.MapService.SERVICE_NAME;
 import static com.hazelcast.test.HazelcastTestSupport.assertTrueEventually;
 import static com.hazelcast.test.HazelcastTestSupport.getAddress;
 import static com.hazelcast.test.HazelcastTestSupport.getNode;
+import static com.hazelcast.test.HazelcastTestSupport.getNodeEngineImpl;
 import static com.hazelcast.test.HazelcastTestSupport.randomMapName;
 import static com.hazelcast.test.HazelcastTestSupport.warmUpPartitions;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 import static org.mockito.Mockito.mock;
 
 @RunWith(HazelcastParallelClassRunner.class)
@@ -56,7 +59,7 @@ public class ClientMapPartitionLostListenerTest {
 
         client.getMap(mapName).addPartitionLostListener(mock(MapPartitionLostListener.class));
 
-        assertRegistrationsSizeEventually(instance, mapName, 1);
+        assertRegistrationEventually(instance, mapName, true);
     }
 
     @Test
@@ -67,10 +70,10 @@ public class ClientMapPartitionLostListenerTest {
         final String mapName = randomMapName();
 
         final String registrationId = client.getMap(mapName).addPartitionLostListener(mock(MapPartitionLostListener.class));
-        assertRegistrationsSizeEventually(instance, mapName, 1);
+        assertRegistrationEventually(instance, mapName, true);
 
         assertTrue(client.getMap(mapName).removePartitionLostListener(registrationId));
-        assertRegistrationsSizeEventually(instance, mapName, 0);
+        assertRegistrationEventually(instance, mapName, false);
     }
 
     @Test
@@ -119,7 +122,9 @@ public class ClientMapPartitionLostListenerTest {
         final ClientConfig clientConfig = new ClientConfig();
         clientConfig.getNetworkConfig().setSmartRouting(false);
         final HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
-        warmUpPartitions(instance1, instance2, client);
+
+        assertMapContainerCreated(instance1, mapName);
+        assertMapContainerCreated(instance2, mapName);
 
         final HazelcastClientInstanceImpl clientInstanceImpl = getHazelcastClientInstanceImpl(client);
         final Address clientOwnerAddress = clientInstanceImpl.getClientClusterService().getOwnerConnectionAddress();
@@ -129,8 +134,8 @@ public class ClientMapPartitionLostListenerTest {
         final EventCollectingMapPartitionLostListener listener = new EventCollectingMapPartitionLostListener(0);
         client.getMap(mapName).addPartitionLostListener(listener);
 
-        assertRegistrationsSizeEventually(instance1, mapName, 1);
-        assertRegistrationsSizeEventually(instance2, mapName, 1);
+        assertRegistrationEventually(instance1, mapName, true);
+        assertRegistrationEventually(instance2, mapName, true);
 
         final MapService mapService = getNode(other).getNodeEngine().getService(SERVICE_NAME);
         final int partitionId = 5;
@@ -139,16 +144,35 @@ public class ClientMapPartitionLostListenerTest {
         assertMapPartitionLostEventEventually(listener, partitionId);
     }
 
-    private void assertRegistrationsSizeEventually(final HazelcastInstance instance, final String mapName, final int size) {
+    private void assertMapContainerCreated(final HazelcastInstance instance, final String mapName) {
         assertTrueEventually(new AssertTask() {
             @Override
             public void run()
                     throws Exception {
+                final MapService mapService = getNodeEngineImpl(instance).getService(MapService.SERVICE_NAME);
+                assertNotNull(mapService.getMapServiceContext().getMapContainer(mapName));
+            }
+        });
+    }
 
+    private void assertRegistrationEventually(final HazelcastInstance instance, final String mapName, final boolean shouldBeRegistered) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
                 final InternalEventService eventService = getNode(instance).getNodeEngine().getEventService();
-                final Collection<EventRegistration> registrations = eventService.getRegistrations(SERVICE_NAME, mapName);
-                assertEquals(size, registrations.size());
 
+                boolean registered = false;
+                for (EventRegistration registration : eventService.getRegistrations(SERVICE_NAME, mapName)) {
+                    if (registration.getFilter() instanceof MapPartitionLostEventFilter) {
+                        registered = true;
+                        break;
+                    }
+                }
+
+                if (shouldBeRegistered != registered) {
+                    fail("shouldBeRegistered: " + shouldBeRegistered + " registered: " + registered);
+                }
             }
         });
     }

--- a/hazelcast-client/src/test/java/com/hazelcast/client/map/ClientMapPartitionLostListenerTest.java
+++ b/hazelcast-client/src/test/java/com/hazelcast/client/map/ClientMapPartitionLostListenerTest.java
@@ -7,6 +7,7 @@ import com.hazelcast.config.Config;
 import com.hazelcast.core.HazelcastInstance;
 import com.hazelcast.map.MapPartitionLostEvent;
 import com.hazelcast.map.MapPartitionLostListenerStressTest.EventCollectingMapPartitionLostListener;
+import com.hazelcast.map.impl.MapPartitionLostEventFilter;
 import com.hazelcast.map.impl.MapService;
 import com.hazelcast.map.listener.MapPartitionLostListener;
 import com.hazelcast.nio.Address;
@@ -21,7 +22,6 @@ import org.junit.Test;
 import org.junit.experimental.categories.Category;
 import org.junit.runner.RunWith;
 
-import java.util.Collection;
 import java.util.List;
 
 import static com.hazelcast.client.impl.ClientTestUtil.getHazelcastClientInstanceImpl;
@@ -29,11 +29,14 @@ import static com.hazelcast.map.impl.MapService.SERVICE_NAME;
 import static com.hazelcast.test.HazelcastTestSupport.assertTrueEventually;
 import static com.hazelcast.test.HazelcastTestSupport.getAddress;
 import static com.hazelcast.test.HazelcastTestSupport.getNode;
+import static com.hazelcast.test.HazelcastTestSupport.getNodeEngineImpl;
 import static com.hazelcast.test.HazelcastTestSupport.randomMapName;
 import static com.hazelcast.test.HazelcastTestSupport.warmUpPartitions;
 import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
 import static org.mockito.Mockito.mock;
 
 @RunWith(HazelcastParallelClassRunner.class)
@@ -56,7 +59,7 @@ public class ClientMapPartitionLostListenerTest {
 
         client.getMap(mapName).addPartitionLostListener(mock(MapPartitionLostListener.class));
 
-        assertRegistrationsSizeEventually(instance, mapName, 1);
+        assertRegistrationEventually(instance, mapName, true);
     }
 
     @Test
@@ -67,10 +70,10 @@ public class ClientMapPartitionLostListenerTest {
         final String mapName = randomMapName();
 
         final String registrationId = client.getMap(mapName).addPartitionLostListener(mock(MapPartitionLostListener.class));
-        assertRegistrationsSizeEventually(instance, mapName, 1);
+        assertRegistrationEventually(instance, mapName, true);
 
         assertTrue(client.getMap(mapName).removePartitionLostListener(registrationId));
-        assertRegistrationsSizeEventually(instance, mapName, 0);
+        assertRegistrationEventually(instance, mapName, false);
     }
 
     @Test
@@ -99,11 +102,9 @@ public class ClientMapPartitionLostListenerTest {
             @Override
             public void run()
                     throws Exception {
-
                 final List<MapPartitionLostEvent> events = listener.getEvents();
                 assertFalse(events.isEmpty());
                 assertEquals(partitionId, events.get(0).getPartitionId());
-
             }
         });
     }
@@ -119,7 +120,9 @@ public class ClientMapPartitionLostListenerTest {
         final ClientConfig clientConfig = new ClientConfig();
         clientConfig.getNetworkConfig().setSmartRouting(false);
         final HazelcastInstance client = hazelcastFactory.newHazelcastClient(clientConfig);
-        warmUpPartitions(instance1, instance2, client);
+
+        assertMapContainerCreated(instance1, mapName);
+        assertMapContainerCreated(instance2, mapName);
 
         final HazelcastClientInstanceImpl clientInstanceImpl = getHazelcastClientInstanceImpl(client);
         final Address clientOwnerAddress = clientInstanceImpl.getClientClusterService().getOwnerConnectionAddress();
@@ -129,8 +132,8 @@ public class ClientMapPartitionLostListenerTest {
         final EventCollectingMapPartitionLostListener listener = new EventCollectingMapPartitionLostListener(0);
         client.getMap(mapName).addPartitionLostListener(listener);
 
-        assertRegistrationsSizeEventually(instance1, mapName, 1);
-        assertRegistrationsSizeEventually(instance2, mapName, 1);
+        assertRegistrationEventually(instance1, mapName, true);
+        assertRegistrationEventually(instance2, mapName, true);
 
         final MapService mapService = getNode(other).getNodeEngine().getService(SERVICE_NAME);
         final int partitionId = 5;
@@ -139,16 +142,35 @@ public class ClientMapPartitionLostListenerTest {
         assertMapPartitionLostEventEventually(listener, partitionId);
     }
 
-    private void assertRegistrationsSizeEventually(final HazelcastInstance instance, final String mapName, final int size) {
+    private void assertMapContainerCreated(final HazelcastInstance instance, final String mapName) {
         assertTrueEventually(new AssertTask() {
             @Override
             public void run()
                     throws Exception {
+                final MapService mapService = getNodeEngineImpl(instance).getService(MapService.SERVICE_NAME);
+                assertNotNull(mapService.getMapServiceContext().getMapContainer(mapName));
+            }
+        });
+    }
 
+    private void assertRegistrationEventually(final HazelcastInstance instance, final String mapName, final boolean shouldBeRegistered) {
+        assertTrueEventually(new AssertTask() {
+            @Override
+            public void run()
+                    throws Exception {
                 final InternalEventService eventService = getNode(instance).getNodeEngine().getEventService();
-                final Collection<EventRegistration> registrations = eventService.getRegistrations(SERVICE_NAME, mapName);
-                assertEquals(size, registrations.size());
 
+                boolean registered = false;
+                for (EventRegistration registration : eventService.getRegistrations(SERVICE_NAME, mapName)) {
+                    if (registration.getFilter() instanceof MapPartitionLostEventFilter) {
+                        registered = true;
+                        break;
+                    }
+                }
+
+                if (shouldBeRegistered != registered) {
+                    fail("shouldBeRegistered: " + shouldBeRegistered + " registered: " + registered);
+                }
             }
         });
     }

